Snap 1.0 support

imalsogreg opened the pull request:
Upgrade snap1.0 connection to websockets Stream, consuming request body as io-stream instead of enumerator.

For testing, snap 1.0 is included as a git submodule. After the snap1.0 release, we'll remove this and use snap 1.0 from hackage.

jaspervdj commented:
Looks amazing! I'm guessing I should wait for the Snap 1.0 release before I merge this?

imalsogreg commented:
Hugely cleaned up by `makeStream`  :)

I'm not sure what makes the most sense for merging. Snap's 1.0 work is in their master branch, so matching that style is one argument for merging now. OTOH I haven't heard many people asking for snap & websockets who are testing 1.0 - so I don't have a strong opinion about waiting to merge.

imalsogreg commented:
Hi Jasper. I cleaned up the submodule cruft, updated the cabal file & changelog, and tested the example  server (on ghc8 only). Ready to merge after you have time to review. Thanks!

jaspervdj commented on src/Network/WebSockets/Snap.hs:
```diff
@@ -120,37 +62,36 @@ runWebSocketsSnap = runWebSocketsSnapWith WS.defaultConnectionOptions
+  rq <- Snap.getRequest
+  Snap.escapeHttp $ \tickle readEnd writeEnd -> do
+    readEnd'  <- Streams.lockingInputStream  readEnd
+    writeEnd' <- Streams.lockingOutputStream writeEnd
```
I am not sure if locking these is necessary.  `websockets` already adds locking in `makeStream`, so I'm worried the double-locking might impact performance a bit.

jaspervdj commented on CHANGELOG:
```diff
@@ -1,3 +1,8 @@
+- 1.0.0.0
```
I would prefer to bump the version to just `0.10.0.0`, regardless of what snap does.

jaspervdj commented:
It looks great, thanks for doing this!  I've added two small notes and I'm happy to merge after that.

imalsogreg commented:
You're welcome. Thanks for the library! I tweaked the patch, adding the changes you mentioned, and did a small sanity check on the example server (two browser tabs issuing different commands), no signals getting crossed or server crashes or anything like that.

jaspervdj commented:
Thanks!
